deployer/src/proxy.rs

```diff
@@ -33,44 +34,58 @@ pub async fn handle(
     });
     span.set_parent(parent_context);
 
-    let host = match req.headers().get(HOST) {
-        Some(host) => host.to_str().unwrap_or_default().to_owned(),
+    let host: FQDN = match req.headers().get(HOST) {
+        Some(host) => host
+            .to_str()
+            .unwrap_or_default()
+            .parse::<FQDN>()
+            .unwrap_or_default()
+            .to_owned(),
         None => {
-            trace!("proxy request has to host header");
+            trace!("proxy request has no host header");
             return Ok(Response::builder()
                 .status(StatusCode::BAD_REQUEST)
                 .body(Body::empty())
                 .unwrap());
         }
     };
 
-    let service = match host.strip_suffix(&fqdn) {
-        Some(service) => service,
+    if host != fqdn {
+        trace!(?host, "proxy won't serve foreign domain");
+        return Ok(Response::builder()
+            .status(StatusCode::BAD_REQUEST)
+            .body(Body::from("this domain is not served by proxy"))
+            .unwrap());
+    }
 
+    // We only have one service per project, and its name coincides
+    // with that of the project
+    let service = match req.headers().get("X-Shuttle-Project") {
```

Comment:
In the future when we have multiple services per project support, what will `X-Shuttle-Project` be? I don't think we plan to determine the service using `gateway`. Ie `deployer` will figure out which service the request is for.

If deployer will eventually determine the service in the future, then it should be safe to remove this code now? Maybe for now get the project name from the `deployer` args :man_shrugging:

Reply:
> In the future when we have multiple services per project support, what will X-Shuttle-Project be? I don't think we plan to determine the service using gateway. Ie deployer will figure out which service the request is for.

Always the name of the project associated with this instance of deployer. Just the same as `X-Shuttle-Account-Name`.

> If deployer will eventually determine the service in the future, then it should be safe to remove this code now? Maybe for now get the project name from the deployer args

I see where you're coming from. But I think using custom headers is a better option than using CLI arguments. If anything, it makes tracing requests easier. Because this works and was tested and we're on a tight timeline I'll leave it for now, but we'll reopen this when we reopen the plan for multi-services support.